### ISSUE TITLE
iptables should be empty when the firewall is disabled

### ISSUE THREAD

**Issue by wwat** [NixOS/nixpkgs#6290]
After a reconfiguration with the firewall disabled, `iptables -L -v` output contains various rules. This is not the current state in NixOS on my system (14.12). 

So, I am expecting output similar to:
```
Chain INPUT (policy ACCEPT 3226K packets, 4462M bytes)
 pkts bytes target     prot opt in     out     source               destination         

Chain FORWARD (policy ACCEPT 0 packets, 0 bytes)
 pkts bytes target     prot opt in     out     source               destination         

Chain OUTPUT (policy ACCEPT 2508K packets, 460M bytes)
 pkts bytes target     prot opt in     out     source               destination 
```

**Comment by wmertens** [comment]
@wwat The NixOS firewall only concerns itself with the nixos chain, and leaves the rest as-is. I rather like this setup, it allows for ad-hoc rule changes like done by Docker and Vagrant.

I'm not sure that the entire firewall ruleset should be managed, if so, I'd prefer it to be behind a flag.

**Comment by wwat** [comment]
I don't understand you reaction; I am arguing that if a user desires that NixOS doesn't touch iptables configuration, that it allows for that. When debugging one should be able to fully turn off systems and I must say that it annoys me that I need to give reasons for trivial features which pretty much every sub-system should have.

**Comment by wmertens** [comment]
@wwat so are you saying that after turning off the firewall, there are nixos-related chains that are not used and do not interfere with the firewall, but they are still there and you want them to be removed?

```
# iptables -L
The program ‘iptables’ is currently not installed. You can install it by typing:
  nix-env -i iptables

[root@localhost:~]# nix-shell -p iptables --command "!!"
nix-shell -p iptables --command "iptables -L"
Chain INPUT (policy ACCEPT)
target     prot opt source               destination         

Chain FORWARD (policy ACCEPT)
target     prot opt source               destination         

Chain OUTPUT (policy ACCEPT)
target     prot opt source               destination         

Chain nixos-fw (0 references)
target     prot opt source               destination         
nixos-fw-accept  all  --  anywhere             anywhere            
nixos-fw-accept  all  --  anywhere             anywhere             ctstate RELATED,ESTABLISHED
nixos-fw-accept  tcp  --  anywhere             anywhere             tcp dpt:ssh
nixos-fw-log-refuse  all  --  anywhere             anywhere            

Chain nixos-fw-accept (3 references)
target     prot opt source               destination         
ACCEPT     all  --  anywhere             anywhere            

Chain nixos-fw-log-refuse (1 references)
target     prot opt source               destination         
LOG        tcp  --  anywhere             anywhere             tcp flags:FIN,SYN,RST,ACK/SYN LOG level info prefix "rejected connection: "
nixos-fw-refuse  all  --  anywhere             anywhere             PKTTYPE != unicast
nixos-fw-refuse  all  --  anywhere             anywhere            

Chain nixos-fw-refuse (2 references)
target     prot opt source               destination         
DROP       all  --  anywhere             anywhere            
```

The firewall-stop script handles turning off an enabled firewall and it only removes the references to the nixos chains and not the chains themselves: https://github.com/NixOS/nixpkgs/blob/364e308a54902e700ea73ddcacfccfe8563fda18/nixos/modules/services/networking/firewall.nix#L181-L197

This behavior was written by @edolstra in https://github.com/NixOS/nixpkgs/commit/3bc3dc3940ba7079ca76341ade31c6c6a7580a9f - I presume he felt it unnecessary to remove inactive chains in the not-so-common event of turning off a firewall.

So basically, if you would like to have these inactive chains removed, please explain why they are a problem and why we should take on the (small) burden of (some) extra code to handle this, IMHO, rather infrequent event.

**Comment by wwat** [comment]
Yes, that's indeed what I asked for. With your explanation (and the fact that I fixed my iptables issue already), there is less of a need for me, but I still consider it bad design. By keeping those nixos specific rules in there, even if they are not active when the user expects the system to go in a zero state, you force the user to learn something about something he/she doesn't particularly care about. 

Another option would be to patch the iptables code to output NixOS specific changes or to change the documentation of the iptables command in order to explain what you just explained above, but the least amount of work would be if disabling the firewall really meant exactly that. 

If a system requires someone to ask a question about how to use it, it's not documented well enough or is not self-discoverable.

**Comment by wmertens** [comment]
So to fully clear iptables in the event that the user disabled the firewall after it was active (the only time this problem occurs), the firewall-stop script should either:
- remove all added chains (error-prone, need to sync with code elsewhere)
- remove all chains named nixos-fw* (might remove user-maintained chains)

Patching the iptables binary to add informational messages is a bridge too far for me... Maintaining patches to upstream code is not trivial. Wrapping it might be an option but when to show the messages? You'd have to discover the current state and based on that provide information.

I'm not happy about any of these options. If you feel strongly about this, please open a PR with proposed code, and we'll definitely consider it, but until then this is a won't fix for me.

PS: Considering that the entire UI of Unix consists of "$ _", it is probably not self-discoverable. If you mess with iptables I like to presume that you know how it works from reading its manpage.

**Comment by wwat** [comment]
Enabling the firewall does X, disabling the firewall should do the inverse of that. 

If there are any other chains, then there should be configuration for that too, and there should be a knob for disabling *everything* which NixOS has configured w.r.t. iptables. 

Doing a hack like removing all chains named nixos-fw* is obviously not going to work.
